consensus/src/experimental/commit_phase.rs

```diff
@@ -0,0 +1,338 @@
+        author: Author,
+    ) -> Self {
+        Self {
+            commit_channel_recv,
+            execution_proxy,
+            local_cache: HashMap::<HashValue, LedgerInfoWithSignatures>::new(),
+            local_queue: VecDeque::<PendingBlocks>::new(),
+            commit_msg_sender,
+            commit_msg_receiver,
+            verifier,
+            safety_rules,
+            author,
+            committed_round: 0,
+        }
+    }
+
+    /// Notified when receiving a commit vote message
+    pub async fn process_commit_vote(&mut self, commit_vote: &CommitVote) -> anyhow::Result<()> {
+        let li = commit_vote.ledger_info();
+
+        if li.commit_info().round() < self.committed_round {
+            return Ok(()); // we ignore the message
+        }
+
+        // verify the signature
+        commit_vote.verify(&self.verifier)?;
+
+        let executed_state_hash = li.commit_info().executed_state_id();
+
+        // add the signature to local_cache
+        match self.local_cache.entry(executed_state_hash) {
+            Entry::Occupied(mut ledger_info_entry) => {
+                let mut_ledger_info_entry = ledger_info_entry.get_mut();
+                mut_ledger_info_entry
+                    .add_signature(commit_vote.author(), commit_vote.signature().clone());
+            }
+            Entry::Vacant(_) => {
+                let mut li_sig = LedgerInfoWithSignatures::new(
+                    li.clone(),
+                    BTreeMap::<AccountAddress, Ed25519Signature>::new(),
+                );
+                li_sig.add_signature(commit_vote.author(), commit_vote.signature().clone());
+                self.local_cache.insert(executed_state_hash, li_sig);
+            }
+        };
+
+        Ok(())
+    }
+
+    pub async fn process_commit_decision(
+        &mut self,
+        commit_decision: CommitDecision,
+    ) -> anyhow::Result<()> {
+        let li = commit_decision.ledger_info();
+
+        if li.ledger_info().commit_info().round() < self.committed_round {
+            return Ok(()); // we ignore the message
+        }
+
+        commit_decision.verify(&self.verifier)?;
+
+        let executed_state_hash = li.ledger_info().commit_info().executed_state_id();
+
+        // TODO: optimization1: probe local_cache first to see if the existing already verifies,
+        // TODO: otherwise we do not make changes.
+        // TODO: optimization2: we can set a bit to indicate this tree of signatures are already verified,
+        // TODO: we do not have to verify it again in the main loop.
+
+        // replace the signature tree directly if it does not verify
+        self.local_cache.insert(executed_state_hash, li.clone());
+
+        Ok(())
+    }
+
+    pub async fn process_local_queue(&mut self) {
+        let mut_local_queue = &mut self.local_queue;
+        let mut_local_cache = &mut self.local_cache;
+        loop {
+            let front = mut_local_queue.front().unwrap();
```

Comment:
the queue can be empty? can use `while let Some(front) = queue.front()`

Reply:
good idea